### ISSUE TITLE
[5000 SigmaUSD] C + Swift(iOS) bindings 

### ISSUE THREAD

**Issue by greenhat** [ergoplatform/sigma-rust#47]
## Scope
Implement an API feature-wise on par with Wasm bindings (ergo-lib-wasm). Basically an API that a wallet would need to be able to send and receive coins and tokens including input box selection, output box creation and transaction building and signing.

## Architecture
The core part is the common/shared source code with JNI (ergo-lib-jni) that should be implemented in ergo-lib-c-core crate.
The generic C bindings should be implemented in ergo-lib-c which should be used in Swift wrapper that should be implemented in ergo-lib-ios.

### C
The main idea is to pass Rust types wrapped in opaque pointers and expose functions to manipulate them on C side.
The crucial part is to free resources on the same side(usually Rust) where they were allocated. For this we need to expose a separate function for every "resource-producing" function that frees a resource passed from C. In case of Address see address_from_testnet and address_delete.

Use the already implemented Address type as an example. 

### Swift
Wraps functions from ergo-lib-c in classes for resource management and exposes methods converting to/from Swift types where necessary.

## Types
Using Wasm bindings as a reference we need the following types(with methods) implemented. For implmentation details see Wasm types and their methods in ergo-lib-wasm. Some method names are adopted for C types. See implemented Address type

Wallet:
from_secrets
sign_transaction

SecretKey:
dlog_from_bytes
to_bytes

ErgoStateContext:
new(from PreHeader)

PreHeader:
from_block_header

BlockHeader:
from_json

BlockHeaders:
from_json

UnsignedTransaction:
id
inputs
data_inputs
output_candidates
to_json
from_json

ErgoBox:
new
box_id
creation_height
tokens
ergo_tree
value
register_value
to_json
from_json

ErgoBoxCandidate:
creation_height
tokens
ergo_tree
value
register_value

Token:
new
id
amount

TokenId:
from_box_id
from_str
to_str

TokenAmount:
from_int64_t
as_int64_t

BoxValue:
from_int64_t
as_int64_t

Transaction:
id
to_json
from_json
inputs
data_inputs
output_candidates
outputs

DataInput:
new
box_id

TxBuilder:
new
set_data_inputs
build
box_selection
data_inputs
output_candidates
current_height
fee_amount
change_address
min_change_value

ErgoBoxCandidateBuilder:
new
set_min_box_value_per_byte
min_box_value_per_byte
set_value
value
calc_box_size_bytes
calc_min_box_value
set_register_value
register_value
delete_register_value
mint_token
add_token
build

BoxSelection:
new
boxes
change

SimpleBoxSelector:
select

ErgoBoxAssetsData:
new
value
tokens

## Collections
The following types represent a simple wrapper for Vec<T>. Feel free to explore the idea to use the single type (pointer+size) since Rust arrays and slices have the same representation in C (arrays).

ErgoBoxCandidates; 
Tokens;
DataInputs;
ErgoBoxAssetsDataList;
SecretKeys;
ErgoBoxes (instead of UnspentBoxes, OutputBoxes, DataInputBoxes)

All of them should implement the following methods:
new
len
get
add

The ErgoBoxes type should implement from_boxes_json method (see Wasm).

**Comment by kettlebell** [comment]
Progress update:

 - I'm up to speed on the coding side and have successfully bound a number of `ergotree_ir::chain::address:Address` methods with unit tests.
 - I've figured out how to build and test the package using only the swift package manager. (see https://github.com/ergoplatform/sigma-rust/pull/468/files#diff-830625053fc44a3b69efb2b846fccb24561bcdb1d18e822a8691066dac1c3ee0R14-R19) It only just occurred to me to see how it's currently done with CI. Should we keep the current build process with `cargo-lipo` and `xcodebuild`?

**Comment by greenhat** [comment]
> Progress update:
> 
>     * I'm up to speed on the coding side and have successfully bound a number of `ergotree_ir::chain::address:Address` methods with unit tests.

Hey! It's looking great!
 
>     * I've figured out how to build and test the package using only the swift package manager. (see https://github.com/ergoplatform/sigma-rust/pull/468/files#diff-830625053fc44a3b69efb2b846fccb24561bcdb1d18e822a8691066dac1c3ee0R14-R19) It only just occurred to me to see how it's currently done with CI. Should we keep the current build process with `cargo-lipo` and `xcodebuild`?

I don't think we should hold on to cargo-lipo since it's not actively maintained anymore - https://github.com/TimNN/cargo-lipo#maintenance-status
Feel free to update ios.yml GA to your workflow. Keep in mind that our goal is to ship Swift lib that will be as close to native iOS development workflow as possible. So for an iOS developer, it does not look like an alien but like any other native Swift lib that they are using.

**Comment by kettlebell** [comment]
> I don't think we should hold on to cargo-lipo since it's not actively maintained anymore - https://github.com/TimNN/cargo-lipo#maintenance-status Feel free to update ios.yml GA to your workflow. Keep in mind that our goal is to ship Swift lib that will be as close to native iOS development workflow as possible. So for an iOS developer, it does not look like an alien but like any other native Swift lib that they are using.

It appears that libs built with Swift Package Manager work out of the box with XCode since version 11. I'll update the workflow. Thanks.

**Comment by kettlebell** [comment]
Hi @greenhat I think I've found an issue when porting over a `wasm` test to `swift`. Looking here: https://github.com/ergoplatform/sigma-rust/blob/88cdfeb6b9e9d86f388e80a5429f59d21ec34f10/bindings/ergo-lib-wasm/tests/test_transaction.js#L190-L192

I believe it should be `const new_tx = new_tx_builder.build();`, which gives an error `Not enough coins(10000100000 nanoERGs are missing)`. I recreated this error on the swift side. Any suggestions on what to do here?

**Comment by greenhat** [comment]
> Hi @greenhat I think I've found an issue when porting over a `wasm` test to `swift`. Looking here:
> 
> https://github.com/ergoplatform/sigma-rust/blob/88cdfeb6b9e9d86f388e80a5429f59d21ec34f10/bindings/ergo-lib-wasm/tests/test_transaction.js#L190-L192
> 
> I believe it should be `const new_tx = new_tx_builder.build();`, which gives an error `Not enough coins(10000100000 nanoERGs are missing)`. I recreated this error on the swift side. Any suggestions on what to do here?

Good catch! Please make and use new output instead `tx_outputs`. Use value for it 10 times lower than in `outbox` value.

**Comment by kettlebell** [comment]
I've confirmed that it now works, but I also needed to change the target balance from `BoxValue.SAFE_USER_MIN()` to the new output + fee.

One other thing I noticed is that the GA configuration for `wasm` doesn't run the tests within the the javascript files, only the single test within `transaction.rs`. I got it working on my side by installing `webpack` and running
```
npm run build
npm run test
```
I'm not too knowledgeable about JS matters, what do you think?

**Comment by kettlebell** [comment]
Back to the issue at hand, the bindings are feature complete and the test suite matches that in `wasm`. Just need to clean up a bit of the code and add documentation.

**Comment by greenhat** [comment]
> I've confirmed that it now works, but I also needed to change the target balance from `BoxValue.SAFE_USER_MIN()` to the new output + fee.
> 
> One other thing I noticed is that the GA configuration for `wasm` doesn't run the tests within the the javascript files, only the single test within `transaction.rs`. I got it working on my side by installing `webpack` and running
> 
> ```
> npm run build
> npm run test
> ```
> 
> I'm not too knowledgeable about JS matters, what do you think?

You're right. For some reason I put JS tests in to a separate js_tests.yml GA workflow. I cannot recall the reason and giving that I cannot think of a use case for ergo-lib-wasm **without** JS we probably should merge wasm_test.yml into js_tests.yml.

**Comment by greenhat** [comment]
> Back to the issue at hand, the bindings are feature complete and the test suite matches that in `wasm`. Just need to clean up a bit of the code and add documentation.

Awesome! You did a fantastic job on this PR! It's a huge piece of work!

**Comment by kettlebell** [comment]
Just want to elaborate on some issues relating to xcode and Apple tooling. First of all I ran it on the latest Xcode version 13.1, and I have an Intel mac.

- Swift Package Manager(SPM) worked well for quickly iterating and testing the code, but despite being an official Apple project it doesn't play well with Xcode at all due to the static lib of `ergo-lib-c`. Unbelievably, there's just no way to point the linker in Xcode to the static lib! Note that SPM works well for GA as it automatically runs our test suite.
- The bindings has a single dependency on https://github.com/SwiftyJSON/SwiftyJSON, which is easy to bring in through SPM. But again, `xcodebuild` doesn't like it, thinking swiftyJSON is another target alongside ErgoLib! So as a quick workaround, I vendored it. It's not the worst, since it's just a single source file. In the future I believe I could use another package manager like Carthage to bring it in cleanly.
- I had to split out the `ergo-lib-c/Config/Release.xcconfig` file into separate ones each for `iphoneos` and `iphonesimulator`. The fix for `iphonesimulator` was super obscure: I had to replace line 3 `VALID_ARCHS = ..." with "EXCLUDED_ARCHS = ...". By the way, we can close issue #465 and we have some hint about why it was a problem!
- After running `xcodebuild` we still need to manually set some settings within Xcode. I've added instructions in the readme which are specific to Intel-based macs.

It's a relief to get it to this point. I look forward to working on the other issues : )
